Add web server implementation

wang-joseph opened the pull request:
Adds the files related to web server implementation to master.

The web server should be able to properly route requests according to the path, cache files for a short period of time (60 seconds) for later retrieval, and returns a failed response if the routing is wrong or the request is handled invalidly.

Also adds a few handlers that will likely be used for this project, each with a small mock-up HTML file. These will be replaced with calls to the templater and services to generate correct pages.

wang-joseph committed:
initialized files

initialized related files according to planning UML

wang-joseph committed:
initialized handlers

I need some of these to create the server so I decided to write a quick initialization

wang-joseph committed:
http messages initial implementation

basic implementation of the http message objects, can be modified later as needed

wang-joseph committed:
basic initialization of files v2

- initialized files according to the new updated completed uml
- added javadocs detailing each file
- added basic implementation for files that don't require a lot of logic

wang-joseph committed:
refactored the connection handler class

- refactored the connection handler class and moved it as an inner class under web server to be able to access server class variables without passing a reference to the web server

wang-joseph committed:
implemented basic web server functionality

- running the web server and connecting to paths function
- using the correctly typed route will properly route the request
- the correct route isn't parsed yet so paths that should work (eg. /problems/1) will fail

wang-joseph committed:
revamped the routing system

- routes are now similar to express.js in terms of matching
- routes now serve as a form of "regex" to match various paths while not needing to hardcode specific identifiers in the path
- routes also route to specific handlers and new handlers will generate new information

wang-joseph committed:
implmemented web cache functionality

- all methods in web cache are now functional
- the web server will search in the cache first before looking for handlers to handle the request
- the web server will save generated html pages in the cache
- the cache is... questionably synchronized. Look into proper cache synchronization later, as Collections.synchronizedSet() breaks cache synchronization

wang-joseph committed:
Create package-info.java

added package-info to the webcache folder

wang-joseph committed:
Implemented query string functionality

- query strings are now seperate from the path and are not considered during routing
- request objects now store a map for query strings for easy retrieval
- the full path can still be used for caching operations
- changed many hashmap occurances with generic map objects

wang-joseph committed:
Made WebLruCache get consistent

- made weblrucache return null instead of an empty string if it cannot find the cached object
- this makes it more consistent with other get operations where null is returned upon failure

wang-joseph committed:
handling of improper request strings

- if an improper status line is given for any reason whatsoever it will return a failed response instead of crashing
- I'm not actually sure if this will ever happen but better safe than sorry

wang-joseph committed:
minor changes in preparation for pr

- deleted the main function
- added some more documentation
- removed unnecessary imports

beepboop271 commented on src/dubjhandlers/HomeHandler.java:
```diff
@@ -0,0 +1,16 @@
+    String body = "<html><head><TITLE>HOME!</TITLE></head><body>This is the home page!</body></html>";
+    Response newResponse = new Response(200, body);
+    newResponse.addHeader("Content-Type", "text/html");
+    newResponse.addHeader("Content-Length", "81");
```
Manually adding headers is quite unnecessary. You can add some static methods in Response, such as Response.notFound( ... ) to generate a 404, or Response.okHtml( ... ) to generate a 200 with the Content-Type and Content-Length headers (and possibly others) set within.

beepboop271 commented on src/dubjhandlers/ProblemHandler.java:
```diff
@@ -0,0 +1,15 @@
+import webserver.RouteTarget;
+import webserver.Response;
+import webserver.Request;
+public class ProblemHandler implements RouteTarget {
```
Missing a blank line.

beepboop271 commented on src/webserver/HttpMessage.java:
```diff
@@ -0,0 +1,214 @@
+import java.util.Map;
+
+/**
+ * This class is the abstract server-side base implementation class for HTTP
```
Your comment lines are wrapped too long, use the formatter.

You can press Ctrl+K+F with the cursor inside a javadoc block to automatically format that block.

beepboop271 commented on src/webserver/HttpMessage.java:
```diff
@@ -0,0 +1,214 @@
+   * A constructor for a new HttpMessage with headers but no body, for invokation
+   * of implementing subclasses.
+   * 
+   * @param headers a {@code HashMap<String, String>} with applicable headers.
```
Capitalize your doc sentences (e.g. `A {@code HashMap...`).

beepboop271 commented on src/webserver/HttpMessage.java:
```diff
@@ -0,0 +1,214 @@
+   * Adds a group of headers to this message's list of headers.
+   * 
+   * @param headers a header-name header-detail map of headers to add.
+   */
+  public void addHeaders(Map<String, String> headers) {
+    this.headers.putAll(headers);
+  }
+
+  /**
+   * Adds a string array of headers to this message's list of headers. Headers in
+   * the array will be parsed, seperated by a colon.
+   * <p>
+   * For example: {@code ["Connection: Keep-Alive", "Accept-Language: en-us"]}
+   * will be parsed into
+   * {@code Connection: "Keep-Alive", Accept-Language: "en-us"}.
+   * <p>
+   * Invalid headers will be thrown out.
+   * 
+   * @param headers a string array of headers.
+   */
+  public void addHeaders(String[] headers) {
+    for (String s : headers) {
+      // seperate the headers info and place into hashmap
+      if (s.indexOf(":") != -1) {
+        String headerName = s.substring(0, s.indexOf(":")).trim();
+        String headerDetails = s.substring(s.indexOf(":")).trim();
+
+        if (!headerName.equals("") && !headerDetails.equals("")) {
+          this.headers.put(headerName, headerDetails);
+        }
+      }
```
Maybe have a `void addHeader(String header)` too. In addition, don't silently ignore things like missing colon or empty key/value. The 400 response code exists for a reason. Abort here, then somewhere in the server can handle the exception with a 400 reply.

wang-joseph replied:
What would `void addHeader(String header)` do? Would it assign some default value to the header, or do some headers support no value inside them?

beepboop271 replied:
It splits a single String and adds it to the headers, as opposed to a method that can only take multiple strings.

beepboop271 commented on src/webserver/HttpMessage.java:
```diff
@@ -0,0 +1,214 @@
+   * As noted by the {@link java.util.Map#get(Object) get method used},
+   * {@code null} will be returned if no key was found. {@code null} does not
+   * necessarily mean that a key does not exist, as the key could have been mapped
+   * to {@code null}. Use {@link #hasHeader(String)} to check if a key exists.
```
You should probably just fail on/ignore null headers when adding them.

beepboop271 commented on src/webserver/HttpMessage.java:
```diff
@@ -0,0 +1,214 @@
+    String headerString = "";
+
+    for (String s : this.headers.keySet()) {
+      headerString += s + ": " + this.headers.get(s) + "\n";
```
Don't repeatedly concatenate within a loop (you can ask Candice how that worked out for her Huffman encoder). Use a StringBuilder. Also, HTTP uses `\r\n` as a line delimiter. In addition, there must be an additional blank line containing a single `\r\n` after the last header line, regardless of whether there is a body or not, Decide where you are going to be adding in this newline, as there should not be two.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+        this.workers.execute(new ConnectionHandler(client));
+      }
+
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
```
This is not a very good catch block. In addition, you don't know if you're catching a failed accept or a failed socket creation.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+   * The characters {@code *}, {@code ?}, and {@code +} are all subsets of their
+   * regex counterparts. {@code *} and {@code +} are specifically used as
+   * wildcards matching anything 0-unlimited and 1-unlimited, respectively. the
+   * hyphen {@code -} and dot {@code .} character are interpreted literally.
```
Very unclear.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+   * <p>
+   * Caution must be placed with routing two strings that match the same path, as
+   * non-deterministic routing resolution will occur. There will be no guarantees
+   * as to which route target will handle the request.
```
It is deterministic, you just can't make any guarantees. You could make it both deterministic and guaranteed to have insertion order based priority with a LinkedHashMap.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+   * // This will only match the /static path, nothing more, nothing less.
+   * server.route("/static", new StaticHandler());
+   * // This will match /problems, /problems/problem1, /problems/problem2, etc.
+   * server.route("/problems/?*", new ProblemHandler());
```
I suggest you look at syntax from the routers of other web servers before continuing with this regex based one, especially given the way you've implemented it (see following comments).

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+      }
+    }
+
+    cleanedRoute += "$";
```
- Again with the bad loop concatenation.
- Creating an exclude set of characters is often less effective than creating a set of characters that are only allowed. For example, there are many other characters that would 'mess up' the regex like parentheses and braces. In addition, again, you should not silently ignore bad input, just reject the route string from the method as a whole. Then, you can add a filter as mentioned at the top to only allow valid HTTP url characters + the characters you need for your router in the route string.
- When given the route `/problems/?*` (as your docs say), your program produces the regex `^/problems/?.*$` which allows all of the following as part of that route:
  - `/problems/problem1`
  - `/problems`
  - `/problems/`
  - `/problems///////////`
  - `/problems/problem1/there/is/no/way/to/match/one/layer/its/always/recursive`
  - `/problemsbutitsnotactuallyproblems`
  - `/problemsnot/notproblem1`
  - etc...
- Judging by the fact that you used regex quite literally, I'll assume you didn't think of glob syntax (what was intended).
  - `/problems/*` should be `^/problems/[^/]+$`
  - `/problems/**/*` should match `/problems/blah/blah/something` (recursive) or maybe just have `/problems/**`, up to you
- Re: the route syntax of other libraries, usually there is a way to capture a specific part of the route, for example in express: `/problems/:problemId` would match `/problems/something` and then provide the text "something" accessible in some field of the request object.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+      if (route.matches(possibleRoute)) {
+        return this.routes.get(possibleRoute);
+      }
+    }
```
You can use Map.values instead of Map.keySet.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+
+      try {
+        this.input = new BufferedReader(new InputStreamReader(this.client.getInputStream()));
+        this.output = new PrintWriter(this.client.getOutputStream());
```
You'll have to be careful with PrintWriter:
- Manually set it to use UTF-8 instead of the system default
- Never use println because HTTP always uses \r\n not the system default

At least you are flushing it right.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+
+            // Since readLine() strips the \n, we can safely use \n as a delimiter for split
+            while (input.ready()) {
+              headerLine += input.readLine() + "\n";
```
Again with the bad loop concatenation.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+     */
+    public void run() {
+      String statusLine = "";
+      String headerLine = "";
```
`headerLines` instead?

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+   * @version 0.0.1
+   * @author Joseph Wang
+   */
+  private class ConnectionHandler implements Runnable {
```
You're forgetting a lot of `this.` in this class. Not sure about your other classes but it's very obvious in this one. Accessing the outer class can't use `this` I think but you should still use `this` for inner class access.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+        }
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
```
Are you sure this is a good way to determine if the request is over? This only works when the reader just happens to be able to get all the data in one go without waiting. There are ways to determine if a request is over by looking at the data, not the stream, especially since (with a keep-alive header) multiple requests can be sent on the same socket connection stream (which you appear to have missed).

In addition, you may want to consider moving some of this into Request as something like knowing when a request is over should probably belong in Request, not ConnectionHandler.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+
+      method = statusTokens[0];
+      fullPath = statusTokens[1];
+      protocol = statusTokens[2];
```
This probably belongs in Request.

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+      body = othersLine.substring(othersLine.lastIndexOf("\n"));
+      // Everything else is a lot of headers
+      othersLine = othersLine.substring(0, othersLine.lastIndexOf("\n"));
+      headers = othersLine.split("\n");
```
This probably belongs in Request. It also doesn't consider the fact that the body can contain newlines. Also, the fact that you've converted the `\r\n`s into `\n`s might get confusing and hard to keep track of, in fact the first thing I thought when I saw this was "are the newlines wrong again?" and not "oh, these are the strings that are converted from \r\n into \n".

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+      response.addHeader("Content-Length", Integer.toString(failedBody.length()));
+
+      return response;
+    }
```
This probably belongs in Response (see the comment made in HomeHandler.java).

beepboop271 commented on src/webserver/WebServer.java:
```diff
@@ -0,0 +1,383 @@
+      Response response = processRecievedStrings(statusLine, headerLine);
+
+      // Output to user and shut down
+      output.println(response);
```
I know I already said to be careful with PrintWriter above but I'll just add another comment (because it looks like you didn't read about HTTP requests too much): you *cannot* add a trailing newline to the body, and there should be a *single* blank line after headers (regardless of whether or not there is a body). I suggest you handle all whitespace carefully in the Request and Response classes so that you purely `print(` in the server classes, never adding whitespace.

beepboop271 commented on src/webserver/Request.java:
```diff
@@ -0,0 +1,223 @@
+      return fullPath;
+    }
+
+    String[] queries = fullPath.substring(fullPath.lastIndexOf("?") + 1).split("&");
```
`startIndex+1`

beepboop271 commented on src/webserver/Request.java:
```diff
@@ -0,0 +1,223 @@
+
+      // Ignore improperly formatted queries
+      if (queryInfo.length == 2) {
+        this.queryStrings.put(queryInfo[0], queryInfo[1]);
```
You need to handle percent encoded characters somewhere.

beepboop271 commented on src/webserver/Response.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    return responseString;
+  }
```
Again, make sure you follow the HTTP spec. `\r\n` for newlines. One blank line after headers. It appears you have the right amount of newlines, but they're not `\r\n`.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+   * Constructs a new WebLruCache, with a capacity of 100.
+   */
+  public WebLruCache() {
+    this(DEFAULT_MAX_CAPACITY);
```
`WebLruCache.DEFAULT_MAX_CAPACITY`

beepboop271 commented on src/webserver/webcache/Node.java:
```diff
@@ -0,0 +1,68 @@
+ * implementation.
+ * <p>
+ * This node class stores pointers to both the previous and next node. If the
+ * pointers do not exist, the values will be {@code null}.
```
Interesting phrasing. They are also references, not pointers. In addition, a reference to null exists, it's just referencing null. Just say something about head and tail having nulls.

beepboop271 commented on src/webserver/webcache/TimedObject.java:
```diff
@@ -0,0 +1,93 @@
+  private T object;
+
+  /**
+   * Constructs a new TimedObject, with the default expiration time of 1 second.
```
There are 60 seconds in 1 second?

beepboop271 commented on src/webserver/webcache/TimedObject.java:
```diff
@@ -0,0 +1,93 @@
+   * 
+   * @param update the new updated object.
+   */
+  public void updateObject(T update) {
```
`public void setObject(T object) {`

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+   * @param maxCapacity the max capacity of this cache.
+   */
+  public WebLruCache(int maxCapacity) {
+    this.maxCapacity = maxCapacity;
```
Missing validation/exception throwing on bad capacity.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+            this.lookup.remove(key);
+            break;
+          }
+        }
```
This is O(n) time, not just O(n) worst case. You could consider using a second map.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+  
+      // Create and insert the new node into the proper areas
+      Node<TimedObject<String>> newNode = new Node<TimedObject<String>>(new TimedObject<String>(item, secondsToLive),
+          null, head);
```
Clearly you are still using the default formatter and not the one I gave you, if it decided this was the best way to break the line.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+   * @param secondsToLive the amount of seconds this item has to live.
+   */
+  public synchronized void putCache(String item, String id, int secondsToLive) {
+    synchronized (lookup) {
```
Every method you do this for halves my sanity. Why did you do this?? The map doesn't even need to be synchronized when there is the ConcurrentHashMap class which supports concurrent access as opposed to synchronized access. The only thing that needs to be synchronized is your list.

beepboop271 commented on src/webserver/webcache/TimedObject.java:
```diff
@@ -0,0 +1,93 @@
+ * @author Joseph Wang
+ * 
+ */
+public class TimedObject<T> {
```
Is there any point to separating this from Node?

wang-joseph replied:
Not really, besides that others can... extend these for their own purposes?

It does make sense to just combine this and Node anyways since this is already in the webcache package and only the webcache uses it.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+    }
+  }
+
+  /**
+   * Updates a cached item.
+   * <p>
+   * The object stored under the existing id will be swapped with the new item. If
+   * the id does not exist, nothing happens.
+   * <p>
+   * This method is similar to calling {@link #putCache(String, String, int)} with
+   * a replacement object, but using this method indicates an update. This cached
+   * object will also retain the existing amount of seconds to live from the old
+   * object.
+   * 
+   * @param id      the id to update.
+   * @param newItem the new item to replace the old item.
+   */
+  public synchronized void updateCache(String id, String newItem) {
+    synchronized (this.lookup) {
+      if (checkCache(id)) {
+        this.lookup.get(id).data.updateObject(newItem);
+      }
+    }
+  }
+
+  /**
+   * Attempts to retrieve a cached object using a specified key.
+   * <p>
+   * If the cached object and key does not exist, {@code null} will be
+   * returned.
+   * 
+   * @param id the id of the cached object to fetch.
+   * @return the cached object, or {@code null}.
+   */
+  public synchronized String getCachedObject(String id) {
+    synchronized (this.lookup) {
+        return this.lookup.get(id).data.getObject();
+    }
+  }
```
Forgot to bump the node to the head, thus making it a FIFO cache and not an LRU cache. Also, more synchronized pain.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+    }
+  }
+
+  /**
+   * Clears the cache and removes all cached objects.
+   */
+  public void clearCache() {
+    this.lookup.clear();
+  }
```
If you were to continue using a HashMap, this should be synchronized too. However, ConcurrentHashMap would not need this.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+   * thread, routinely cleaning up the cache.
+   */
+  private void initializeCleanupThread() {
+    Thread cleanupThread = new Thread(new Runnable() {
```
No anonymous classes please.

beepboop271 commented on src/webserver/webcache/WebLruCache.java:
```diff
@@ -0,0 +1,195 @@
+  /**
+   * Initializes the clean-up service, which will run on a low priority background
+   * thread, routinely cleaning up the cache.
+   */
+  private void initializeCleanupThread() {
+    Thread cleanupThread = new Thread(new Runnable() {
+      public synchronized void run() {
+        while (runCleanup) {
+          synchronized (lookup) {
+            for (String toCheck : lookup.keySet()) {
+              if (lookup.get(toCheck).data.isExpired()) {
+                // Remove expired cache objects
+                Node<TimedObject<String>> toRemove = lookup.get(toCheck);
+                lookup.remove(toCheck);
+  
+                if (toRemove.next != null) {
+                  toRemove.next.prev = toRemove.prev;
+                }
+  
+                if (toRemove.prev != null) {
+                  toRemove.prev.next = toRemove.next;
+                }
+  
+                if (toRemove == head) {
+                  head = toRemove.next;
+                }
+  
+                if (toRemove == tail) {
+                  tail = toRemove.prev;
+                }
+              }
+            }
+          }
+        }
+      }
+    });
+
+    cleanupThread.setPriority(Thread.MIN_PRIORITY);
+    cleanupThread.start();
+  }
```
Do you think that marking this thread as MIN_PRIORITY magically makes it pause for a few seconds between each loop? You're effectively locking the map over and over in an infinite loop, increasing the time the other methods need to wait for a chance to lock the map because now there is this loop trying to lock the map every chance it gets. Even after you change to a ConcurrentHashMap and don't need to lock to read the map, please add a 10 second delay or something longer.

beepboop271 reviewed (changes requested):
In general:
- Missing usage of `this` + not using the linter/formatter I gave
- Some shuffling of request/response handling might be good
- Clearly lacking knowledge on how an HTTP request is formatted
- Cache is just weird

wang-joseph committed:
fixed minor formatting and javadoc issues

- cleaned some javadoc in response to review

wang-joseph committed:
fixed cache synchronization and issues

- reflected review changes on the class
- (hopefully) properly inplemented synchronization instead of flailing around like a dead fish and slapping synchronized on everyting
- combined node and timed object
- hopefully this no longer drains kevin's sanity

wang-joseph committed:
added static methods to Response, handled percent encoding and bad handlers

- Response now has static creation methods for generic http responses
- minor formatting changes
- Exception throwing for bad requests and headers
- query strings are decoded, and the path is also decoded (unsure if the path needs to be outside of a form)
- small changes according to review

wang-joseph committed:
request parsing and handling in proper files

- added RequestBuilder to assist in constructing requests
- added exceptions to ensure that parsing can be caught and handled appropriately
- changed main server logic

wang-joseph committed:
fix broken server changes as a result of RequestBuilder

- request accepting and constructing work (for now)

shari09 committed:
router system finished

- regex path router
- read WebServer#route for more info

wang-joseph committed:
connections respect Connection header, minor updates

- now checks the map of headers and makes sure they're "valid" instead of blindly assuming they are
- added docs, this. etc.
- changed cached thread pool to a thread pool, with a minimum and maximum

wang-joseph committed:
Merge branch 'web-server' of https://github.com/beepboop271/duber-judge into web-server

wang-joseph committed:
Fixed minor merging issues

shari09 committed:
fixed router

I accidentally appended another '/' at the end of the regex path.

wang-joseph committed:
Added cookie and cache responses, functionality, minor changes

- cookies have their own creation methods from Response class
- the cache now pays attention to the Cache-Control header to determine whether to cache a page or not. NOTE THAT caching (specifically cache verification) has not been fully implemented yet
- minor changes around general structure

wang-joseph committed:
removed Transfer-Encoding processing, small formatting changes

- removed the Transfer-Encoded header and handler
- general documentation changes and formatting changes to prepare for pr review

wang-joseph committed:
fixed query string bug

wang-joseph commented:
Made a lot of changes in response to feedback: 
- Properly installed and configured formatted and re-formatted every file
- Brushed up on HTTP protocol (didn't google enough previously) to hopefully make these changes more HTTP 1.1-compliant
- (Should have) fixed cache, made into a proper LRU cache instead of a FIFO cache
- Refactored code to have certain parts in certain, more relevant files

**NOTE A major issue occurs with POST requests, where `readline()` never ends because HTTP requests do not require a trailing CRLF. This will be fixed asap**

As for specifics:

**Cache**
- Changed to LRU cache
- Fixed synchronization badness by swapping to `ConcurrentHashMap`, using an object lock for the doubly linked list, and not starving the access threads from my removal background thread
- Combined `Node` and `TimedObject` into one `TimedNode`

**Server**
- Requests are processed and built using a `RequestBuilder`, which will look at the headers to determine a body and can indicate when a request is finished
- Connection persists as per HTTP/1.1 standards for a set amount of time that can be changed with the appropriate headers
- Generic response formation moved to the `Response` class as static methods
- Routing reworked (by @shari09) and can now capture wildcards as params stored on the `Request` object

**HTTP Messages**
- Generic responses can be created with the appropriate call to `Response` static methods, with status code and appropriate headers
- Requests now contain decoded query strings and params from wildcards from the routing
- Checked exceptions are thrown on bad syntax or headers, allowing server to catch and respond to client with a 400

Note/todo:
- Creating a separate function for parsing percent-encoding may want to be implemented — the current implementation should be okay for most cases but using the percent-encoding for = or & will potentially ruin the query string... should not be that big of a deal as handlers should be prepared to handle invalid query strings if they need them, but if we have time, rewriting the decoding code wouldn't hurt
- Validating cache has not been implemented yet... do we need to?
- A bit more header testing wouldn't hurt

wang-joseph committed:
rewrote RequestBuilder to work with posts

- stopped using bufferedreader because nextline() breaks with post bodies as bodies are not required to have a CRLF
- refactored RequestBuilder to be better designed

wang-joseph committed:
fixed bad body comparison

wang-joseph commented:
Fixed the previously mentioned issue, and this PR is now ready for re-review.

wang-joseph committed:
added generic responses for redirects

- redirect responses now exist

wang-joseph committed:
introduced endresource property in request

- endresource property should be able to help in handling

shari09 committed:
added combined named param + capture group match

- you can now do /:greeting(hi|hello) and the capture group match of either 'hi' or 'hello' can be retrieved using .getParam('greeting')
- also updated router docs

wang-joseph committed:
added more Response functionality

- added HEAD method support for generic Responses
- added Forbidden and Method Not Allowed support

wang-joseph committed:
fixed MethodNotAllowed response

wang-joseph committed:
introduced cookie processing

- cookies are now conveniently stored and parsed
- you can now set multiple cookies
- requests use cookies from the stream, response uses it in set-cookie

wang-joseph committed:
updated Responses, include more HEAD functionality

- 201 Created was added
- web server cache now only gets the headers of the cached body if the request method is HEAD

wang-joseph committed:
http sending files is functional

- rewrote message code to support byte arrays as body
- cache should only be storing html files
- output uses bufferedoutputreader now

wang-joseph committed:
fixed web server caching issue

- initialized parameters before checking it

wang-joseph committed:
added TemporaryRedirect generic response method

wang-joseph committed:
added body-to-string functionality in request

- most if not all requests should just be strings anyways so we can simply get back as string

wang-joseph committed:
added 500 Internal Server Error generic constructor

also added a missing import
